File: 137_Climbing Stairs.py
"""
137. Climbing Stairs
Dynamic programming / Fibonacci sequence for step counts.
"""
def climb_stairs(n: int) -> int:
    if n <= 2: return n
    a, b = 1, 2
    for _ in range(3, n+1):
        a, b = b, a + b
    return b

if __name__ == '__main__':
    print('climbStairs(2) =', climb_stairs(2))
    print('climbStairs(3) =', climb_stairs(3))

# Time Complexity: O(n)
# Space Complexity: O(1)

File: 137_Climbing Stairs.cpp
/*
137. Climbing Stairs
Return number of ways to climb n stairs taking 1 or 2 steps (Fibonacci).
*/
#include <bits/stdc++.h>
using namespace std;

int climbStairs(int n) {
    if (n <= 2) return n;
    int a = 1, b = 2;
    for (int i = 3; i <= n; ++i) {
        int c = a + b;
        a = b; b = c;
    }
    return b;
}

int main() {
    cout << "climbStairs(2) = " << climbStairs(2) << "\n"; // 2
    cout << "climbStairs(3) = " << climbStairs(3) << "\n"; // 3
    return 0;
}

// Time Complexity: O(n)
// Space Complexity: O(1)
